ecs-agent/api/ecs/client/ecs_client.go

```diff
@@ -0,0 +1,802 @@
+	"github.com/aws/amazon-ecs-agent/ecs-agent/ec2"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/httpclient"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger/field"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/utils"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/utils/retry"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/docker/docker/pkg/system"
+)
+
+const (
+	ecsMaxImageDigestLength     = 255
+	ecsMaxContainerReasonLength = 255
+	ecsMaxTaskReasonLength      = 1024
+	ecsMaxRuntimeIDLength       = 255
+	defaultPollEndpointCacheTTL = 12 * time.Hour
+	azAttrName                  = "ecs.availability-zone"
+	cpuArchAttrName             = "ecs.cpu-architecture"
+	osTypeAttrName              = "ecs.os-type"
+	osFamilyAttrName            = "ecs.os-family"
+	// RoundtripTimeout should only time out after dial and TLS handshake timeouts have elapsed.
+	// Add additional 2 seconds to the sum of these 2 timeouts to be extra sure of this.
+	RoundtripTimeout = httpclient.DefaultDialTimeout + httpclient.DefaultTLSHandshakeTimeout + 2*time.Second
+	// Below constants are used for SetInstanceIdentity retry with exponential backoff.
+	setInstanceIdRetryTimeOut         = 30 * time.Second
+	setInstanceIdRetryBackoffMin      = 100 * time.Millisecond
+	setInstanceIdRetryBackoffMax      = 5 * time.Second
+	setInstanceIdRetryBackoffJitter   = 0.2
+	setInstanceIdRetryBackoffMultiple = 2
+)
+
+// ecsClient implements ECSClient interface.
+type ecsClient struct {
+	credentialsProvider              *credentials.Credentials
+	configAccessor                   config.AgentConfigAccessor
+	standardClient                   ecs.ECSStandardSDK
+	submitStateChangeClient          ecs.ECSSubmitStateSDK
+	ec2metadata                      ec2.EC2MetadataClient
+	httpClient                       *http.Client
+	pollEndpointCache                async.TTLCache
+	isFIPSDetected                   bool
+	shouldExcludeIPv6PortBinding     bool
+	sascCustomRetryBackoff           func(func() error) error
+	stscAttachmentCustomRetryBackoff func(func() error) error
+}
+
+// NewECSClient creates a new ECSClient interface object.
+func NewECSClient(
+	credentialsProvider *credentials.Credentials,
+	configAccessor config.AgentConfigAccessor,
+	ec2MetadataClient ec2.EC2MetadataClient,
+	agentVer string,
+	osType string,
+	options ...ECSClientOption) (ecs.ECSClient, error) {
+
+	client := &ecsClient{
+		credentialsProvider: credentialsProvider,
+		configAccessor:      configAccessor,
+		ec2metadata:         ec2MetadataClient,
+		httpClient:          httpclient.New(RoundtripTimeout, configAccessor.AcceptInsecureCert(), agentVer, osType),
+		pollEndpointCache:   async.NewTTLCache(&async.TTL{Duration: defaultPollEndpointCacheTTL}),
```

Comment:
We are setting the default as 12 hours. What happens when the containerd agent does not set any TTL? Currently, we do not expire our entries. We should set the TTL as nil by default and individual agents should set their TTL through the optional api.

Reply:
Currently in agent module we default to 12 hours. Any calling Agent can override this with a different TTL (including nil TTL) through the `WithDiscoverPollEndpointCacheTTL` option. IMO it's better to stick with the current default that we have and override wherever necessary. In most cases, the point of using a TTL cache is for the TTL , so using a TTL cache with nil TTL seems a bit weird to me for the default case (i.e, using a nil TTL is more of an exception rather than the norm).

Reply:
Synced offline, we will keep a default TTL for the client to use for its TTL cache.